### PR TITLE
HDDS-2131. Optimize replication type and creation time calculation in S3 MPU list call.

### PR DESCRIPTION
**Pull request by swagle**
## What changes were proposed in this pull request?
Optimize listMultiPartUpload to not read from openKeyTable.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2131

## How was this patch tested?
Ran related unit tests.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 1]
```diff
@@ -1323,29 +1323,36 @@ public OmMultipartUploadList listMultipartUploads(String volumeName,
...
-                  Instant.ofEpochMilli(omKeyInfo.getCreationTime()));
-
-              upload.setReplicationType(omKeyInfo.getType());
-              upload.setReplicationFactor(omKeyInfo.getFactor());
+                  Instant.ofEpochMilli(multipartKeyInfo.getCreationTime()));
+
+              TreeMap<Integer, PartKeyInfo> partKeyInfoMap =
+                  multipartKeyInfo.getPartKeyInfoMap();
+              if (!partKeyInfoMap.isEmpty()) {
+                PartKeyInfo partKeyInfo =
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: If there are no parts, if partKeyInfoMap.isEmpty()?

Comment (swagle) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: Hi @bharatviswa504 not sure I understand, the map is not empty we get the type and factor from the first part. isEmpty() => size == 0

**Comment by swagle** [comment]
> And also this is missing S3 end change to use the replication type and replication factor and show them in S3StorageType format.
> 
> Need to use this class for conversion. https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3StorageType.java

Thanks for pointing this one out, where does this change go?
Do you mean we should add S3StorageType to org.apache.hadoop.ozone.om.response.s3.multipart.S3InitiateMultipartUploadResponse#S3InitiateMultipartUploadResponse

**Comment by bharatviswa504** [comment]
> > And also this is missing S3 end change to use the replication type and replication factor and show them in S3StorageType format.
> > Need to use this class for conversion. https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3StorageType.java
> 
> Thanks for pointing this one out, where does this change go?
> Do you mean we should add S3StorageType to org.apache.hadoop.ozone.om.response.s3.multipart.S3InitiateMultipartUploadResponse#S3InitiateMultipartUploadResponse

This was already taken care of. Sorry I thought this is missing.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: If the MPU uploaded does not have any parts then I don't see that logic here.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 2]
```diff
@@ -893,17 +893,18 @@ private OmMultipartInfo createMultipartInfo(OmKeyArgs keyArgs,
       // Not checking if there is an already key for this in the keyTable, as
       // during final complete multipart upload we take care of this.
 
-
+      long currentTime = Time.now();
       Map<Integer, PartKeyInfo> partKeyInfoMap = new HashMap<>();
       OmMultipartKeyInfo multipartKeyInfo = new OmMultipartKeyInfo(
-          multipartUploadID, partKeyInfoMap);
+          multipartUploadID, currentTime, keyArgs.getType(),
```
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: This change needs to be done here.https://github.com/bharatviswa504/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/multipart/S3InitiateMultipartUploadRequest.java#L146

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: For write requests old code is not used anymore.

Comment (swagle) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: Hi @bharatviswa504, I had already changed this in my previous patch.

**Comment by swagle** [comment]
Hi @bharatviswa504 the tests are failing since we do not have RATIS-707 fix, it is unrelated. Do you want me to rebase and push?

**Comment by bharatviswa504** [comment]
Thank You @swagle for the verification and for the contribution and @anuengineer for the review.

**Pull request merged**